flink/v1.19/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/TriggerManager.java

```diff
@@ -0,0 +1,296 @@
+  public void open(Configuration parameters) throws Exception {
+    this.rateLimiterTriggeredCounter =
+        getRuntimeContext()
+            .getMetricGroup()
+            .addGroup(
+                TableMaintenanceMetrics.GROUP_KEY, TableMaintenanceMetrics.GROUP_VALUE_DEFAULT)
+            .counter(TableMaintenanceMetrics.RATE_LIMITER_TRIGGERED);
+    this.concurrentRunTriggeredCounter =
+        getRuntimeContext()
+            .getMetricGroup()
+            .addGroup(
+                TableMaintenanceMetrics.GROUP_KEY, TableMaintenanceMetrics.GROUP_VALUE_DEFAULT)
+            .counter(TableMaintenanceMetrics.CONCURRENT_RUN_TRIGGERED);
+    this.nothingToTriggerCounter =
+        getRuntimeContext()
+            .getMetricGroup()
+            .addGroup(
+                TableMaintenanceMetrics.GROUP_KEY, TableMaintenanceMetrics.GROUP_VALUE_DEFAULT)
+            .counter(TableMaintenanceMetrics.NOTHING_TO_TRIGGER);
+    this.triggerCounters =
+        taskNames.stream()
+            .map(
+                name ->
+                    getRuntimeContext()
+                        .getMetricGroup()
+                        .addGroup(TableMaintenanceMetrics.GROUP_KEY, name)
+                        .counter(TableMaintenanceMetrics.TRIGGERED))
+            .collect(Collectors.toList());
+
+    this.nextEvaluationTime =
+        getRuntimeContext()
+            .getState(new ValueStateDescriptor<>("triggerManagerNextTriggerTime", Types.LONG));
+    this.accumulatedChanges =
+        getRuntimeContext()
+            .getListState(
+                new ListStateDescriptor<>(
+                    "triggerManagerAccumulatedChange", TypeInformation.of(TableChange.class)));
+    this.lastTriggerTimes =
+        getRuntimeContext()
+            .getListState(new ListStateDescriptor<>("triggerManagerLastTriggerTime", Types.LONG));
+
+    tableLoader.open();
+  }
+
+  @Override
+  public void snapshotState(FunctionSnapshotContext context) throws Exception {
+    // Do nothing
+  }
+
+  @Override
+  public void initializeState(FunctionInitializationContext context) {
+    LOG.info("Initializing state restored: {}", context.isRestored());
+    this.lock = lockFactory.createLock();
+    this.recoveryLock = lockFactory.createRecoveryLock();
+    if (context.isRestored()) {
+      isCleanUp = true;
+    }
+  }
+
+  @Override
+  public void processElement(TableChange change, Context ctx, Collector<Trigger> out)
+      throws Exception {
+    init(out, ctx.timerService());
+
+    long current = ctx.timerService().currentProcessingTime();
+    Long nextTime = nextEvaluationTime.value();
+
+    // Add the new changes to the already accumulated ones
+    List<TableChange> accumulated = Lists.newArrayList(accumulatedChanges.get());
+    accumulated.forEach(tableChange -> tableChange.merge(change));
+    accumulatedChanges.update(accumulated);
+
+    if (nextTime == null) {
+      checkAndFire(current, ctx.timerService(), out);
+    } else {
+      LOG.info(
```

Comment:
That would lead race conditions when the `processElement` and the `onTimer` method calls "should" happen at the same time. If `nextTime` is set, we can be sure that the `onTimer` will be called when the time is right. The check happens there, and we can avoid the race condition.

Reply:
didn't quite understand the race condition part

Reply:
If we trigger the `checkAndFire` when the `nextTime` is larger then the `current`, then we will have a trigger by this new `checkAndFire`, and also the timer could be called to. It is better to avoid this situation altogether by relying on the timer to trigger when it is needed.